finesse/hardware/tc4820.py

```diff
@@ -0,0 +1,212 @@
+"""This module provides an interface to TC4820 temperature controllers.
+
+Decimal numbers are used for values sent to and read from the device as the values are
+base-10 and using floats could cause rounding errors.
+"""
+import logging
+from decimal import Decimal
+from typing import Any
+
+from serial import Serial, SerialException
+
+
+class MalformedMessageError(Exception):
+    """Raised when a message sent or received was malformed."""
+
+
+class TC4820:
+    """An interface for TC4820 temperature controllers."""
+
+    MAX_POWER = 511
+    """The maximum value for the power property."""
+
+    def __init__(self, serial: Serial, max_attempts: int = 3) -> None:
+        """Create a new TC4820 from an existing serial device.
+
+        Args:
+            serial: Serial device
+            max_attempts: Maximum number of attempts for requests
+        """
+        if max_attempts < 1:
+            raise ValueError("max_attempts must be at least 1")
+
+        self.serial = serial
+        self.max_attempts = max_attempts
+
+    @staticmethod
+    def create(
+        port: str,
+        baudrate: int = 115200,
+        timeout: float = 1.0,
+        max_attempts: int = 3,
+        *serial_args: Any,
+        **serial_kwargs: Any,
+    ) -> "TC4820":
+        """Create a new TC4820.
+
+        Args:
+            port: Serial port name
+            baudrate: Serial port baudrate
+            timeout: How long to wait for read/write operation
+            max_attempts: Maximum number of attempts for requests
+            serial_args: Extra arguments to Serial constructor
+            serial_kwargs: Extra keyword arguments to Serial constructor
+        """
+        # If the user hasn't specified an explicit timeout for write operations, then
+        # use the same as for read operations
+        if "write_timeout" not in serial_kwargs:
+            serial_kwargs["write_timeout"] = timeout
+
+        serial = Serial(port, baudrate, *serial_args, timeout=timeout, **serial_kwargs)
+
+        return TC4820(serial, max_attempts)
+
+    def read(self) -> int:
+        """Read a message from the TC4820 and decode the number as a signed integer.
+
+        Raises:
+            MalformedMessageError: The read message was malformed or the device is
+                                   complaining that our message was malformed
+            SerialException: An error occurred while reading the device
+        """
+        message_bytes = self.serial.read_until(b"^", size=8)
+
+        # Don't handle decoding errors, because these will be caught by bytes.fromhex()
+        # below
+        message = message_bytes.decode("ascii", errors="replace")
+
+        if len(message) != 8 or message[0] != "*" or message[-1] != "^":
+            raise MalformedMessageError("Malformed message received: {message}")
```

Comment:
I would add an explanation in the docstring about what the correctly formatted message should look like. This will simplify life to future developers when trying to figure out why you are making all these checks. The same applies to the checks below.